src/utils/utils.ts

```diff
@@ -0,0 +1,525 @@
+  }
+  if ('result' in message) {
+    throw new rpcErrors.ErrorRPCParse('`result` property must not be defined');
+  }
+  if (!('error' in message)) {
+    throw new rpcErrors.ErrorRPCParse('`error` property must be defined');
+  }
+  parseJSONRPCError(message.error);
+  if (!('id' in message)) {
+    throw new rpcErrors.ErrorRPCParse('`id` property must be defined');
+  }
+  if (
+    typeof message.id !== 'string' &&
+    typeof message.id !== 'number' &&
+    message.id !== null
+  ) {
+    throw new rpcErrors.ErrorRPCParse(
+      '`id` property must be a string, number or null',
+    );
+  }
+  return message as JSONRPCResponseError;
+}
+
+function parseJSONRPCError(message: unknown): JSONRPCError {
+  if (!isObject(message)) {
+    throw new rpcErrors.ErrorRPCParse('must be a JSON POJO');
+  }
+  if (!('code' in message)) {
+    throw new rpcErrors.ErrorRPCParse('`code` property must be defined');
+  }
+  if (typeof message.code !== 'number') {
+    throw new rpcErrors.ErrorRPCParse('`code` property must be a number');
+  }
+  if (!('message' in message)) {
+    throw new rpcErrors.ErrorRPCParse('`message` property must be defined');
+  }
+  if (typeof message.message !== 'string') {
+    throw new rpcErrors.ErrorRPCParse('`message` property must be a string');
+  }
+  // If ('data' in message && !utils.isObject(message.data)) {
+  //   throw new rpcErrors.ErrorRPCParse('`data` property must be a POJO');
+  // }
+  return message as JSONRPCError;
+}
+
+function parseJSONRPCResponse<T extends JSONValue>(
+  message: unknown,
+): JSONRPCResponse<T> {
+  if (!isObject(message)) {
+    throw new rpcErrors.ErrorRPCParse('must be a JSON POJO');
+  }
+  try {
+    return parseJSONRPCResponseResult(message);
+  } catch (e) {
+    // Do nothing
+  }
+  try {
+    return parseJSONRPCResponseError(message);
+  } catch (e) {
+    // Do nothing
+  }
+  throw new rpcErrors.ErrorRPCParse(
+    'structure did not match a `JSONRPCResponse`',
+  );
+}
+
+function parseJSONRPCMessage<T extends JSONValue>(
+  message: unknown,
+): JSONRPCMessage<T> {
+  if (!isObject(message)) {
+    throw new rpcErrors.ErrorRPCParse('must be a JSON POJO');
+  }
+  if (!('jsonrpc' in message)) {
+    throw new rpcErrors.ErrorRPCParse('`jsonrpc` property must be defined');
+  }
+  if (message.jsonrpc !== '2.0') {
+    throw new rpcErrors.ErrorRPCParse(
+      '`jsonrpc` property must be a string of "2.0"',
+    );
+  }
+  try {
+    return parseJSONRPCRequest(message);
+  } catch {
+    // Do nothing
+  }
+  try {
+    return parseJSONRPCResponse(message);
+  } catch {
+    // Do nothing
+  }
+  throw new rpcErrors.ErrorRPCParse(
+    'Message structure did not match a `JSONRPCMessage`',
+  );
+}
+
+/**
+ * Replacer function for serialising errors over RPC (used by `JSON.stringify`
+ * in `fromError`)
+ * Polykey errors are handled by their inbuilt `toJSON` method , so this only
+ * serialises other errors
+ */
+const createReplacer = () => {
```

Comment:
i'm not sure that if the state being encapsulated within an IIFE is the best way to do this, i would imagine it to work better as class

thoughts @CMCDragonkai ?

Reply:
Not sure but I did notice this utils/utils.ts should be just utils.ts. No need for subdirectory.